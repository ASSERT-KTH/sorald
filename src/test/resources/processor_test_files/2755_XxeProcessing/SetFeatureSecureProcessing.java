import org.w3c.dom.Document;
import org.xml.sax.InputSource;

import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;
import java.io.StringWriter;

public class SetFeatureSecureProcessing {
    public static String transform(String xslt, String xml) throws TransformerException {
        TransformerFactory transformerFactory = TransformerFactory.newInstance();  // Noncompliant
        transformerFactory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
        Transformer transformer = transformerFactory.newTransformer(new StreamSource(xslt));

        StringWriter writer = new StringWriter();
        transformer.transform(new StreamSource(xml), new StreamResult(writer));
        return writer.toString();
    }

    public static Document parse(String xmlFile) throws Exception {
        DocumentBuilderFactory df = DocumentBuilderFactory.newInstance();  // Noncompliant
        df.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
        DocumentBuilder builder = df.newDocumentBuilder();
        return builder.parse(new InputSource(xmlFile));
    }
}
