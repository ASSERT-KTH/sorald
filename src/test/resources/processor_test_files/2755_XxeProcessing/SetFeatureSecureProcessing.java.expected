import org.w3c.dom.Document;
import org.xml.sax.InputSource;

import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;
import java.io.StringWriter;

public class SetFeatureSecureProcessing {
    public static String transform(String xslt, String xml) throws TransformerException {
        TransformerFactory transformerFactory = createTransformerFactory();
        Transformer transformer = transformerFactory.newTransformer(new StreamSource(xslt));

        StringWriter writer = new StringWriter();
        transformer.transform(new StreamSource(xml), new StreamResult(writer));
        return writer.toString();
    }

    public static Document parse(String xmlFile) throws Exception {
        DocumentBuilderFactory df = createDocumentBuilderFactory();  // Noncompliant
        DocumentBuilder builder = df.newDocumentBuilder();
        return builder.parse(new InputSource(xmlFile));
    }

    private static TransformerFactory createTransformerFactory() throws TransformerConfigurationException {
        TransformerFactory factory = TransformerFactory.newInstance();
        factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "");
        return factory;
    }

    private static DocumentBuilderFactory createDocumentBuilderFactory() throws ParserConfigurationException {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
        return factory;
    }
}
